rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -0,0 +1,225 @@
+        unsafe {
+            XmfVpxImage_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+impl From<VpxCodec> for XmfVpxCodecType {
+    fn from(codec: VpxCodec) -> Self {
+        match codec {
+            VpxCodec::VP8 => XmfVpxCodecType::VP8,
+            VpxCodec::VP9 => XmfVpxCodecType::VP9,
+        }
+    }
+}
+
+pub fn is_key_frame(buffer: &[u8]) -> bool {
+    if buffer.is_empty() {
+        return false;
+    }
+    buffer[0] & 0x1 == 0
+}
+
+pub struct VpxPacket {
+    pub(crate) ptr: *const XmfVpxPacket,
+}
+
+impl VpxPacket {
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn kind(&self) -> XmfVpxPacketKind {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_GetKind(self.ptr.cast_mut()) }
+    }
+
+    /// # Safety
+    ///
+    /// Obtain the deep copy of the frame from the packet, must be used before any call to VpxEncoder method.
+    /// The frame produced by this method is always valid.
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn frame(&self) -> Option<VpxFrame> {
+        // Safety: it's up to the caller to ensure the packet is valid, the frame will remain valid until the frame itself is destroyed.
+        let frame_ptr = unsafe { XmfVpxPacket_GetFrame(self.ptr) };
+        if frame_ptr.is_null() {
+            None
+        } else {
+            VpxFrame::new(frame_ptr).ok()
+        }
+    }
+
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn is_empty(&self) -> bool {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_IsEmpty(self.ptr) }
+    }
+}
+
+impl Drop for VpxPacket {
+    fn drop(&mut self) {
+        // Safety: Ensure that the pointer is valid and has not been freed before calling the drop function.
+        unsafe {
+            XmfVpxPacket_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+pub struct VpxFrame {
+    ptr: *const XmfVpxFrame,
+}
+
+impl Drop for VpxFrame {
+    fn drop(&mut self) {
+        // Safety: see Safety Note in VpxFrame::new, this is safe to call.
+        unsafe {
+            XmfVpxFrame_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+impl VpxFrame {
+    // Safety Note: the XmfVpxFrame is a deep copy from the vpx packet, so it is safe to use the pointer until the frame is destroyed.
+    pub fn new(ptr: *const XmfVpxFrame) -> Result<Self, VpxError> {
```

Comment:
issue: A "safety note" is not a proof. The argument is also very weak. In this case, there is absolutely nothing stopping anyone from providing a pointer to a `XmfVpxFrame` which was already given to another `VpxFrame`. In this case, you’ll have a double-free. It’s even possible to pass in a garbage pointer using safe Rust only.

```rust
fn foo(ptr: *const u8) { }

let ptr = 19239 as *const u8;
foo(ptr);
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3d5ae4298f49be9c20607e6af473f8b9

Either way, you end up with corrupted memory. This is only safe if you do nothing with the pointer, which is not the case: this function should be `unsafe`. But I recommend that you delete the `new`  function which is not intended to be used by the consumer of the library anyway. Instead build the `VpxFrame` directly from the `VpxPacket::frame` function. You’ll also save you from checking for null twice in a row.

Reply:
you are right, this should be constructed by us only!

Reply:
updated